### PR TITLE
fix(color-contrast): handle text that is outside `overflow: hidden` ancestor

### PR DESCRIPTION
**Pull request by straker**
Closes: #4253

### CODE REVIEW

**Review by WilcoFiers — DISMISSED** [review]

File: lib/commons/dom/get-visible-child-text-rects.js [hunk 1]
```diff
@@ -39,8 +39,14 @@ const getVisibleChildTextRects = memoize(
...
      * @see https://github.com/dequelabs/axe-core/issues/2681
+     *
+     * also need to resize the nodeRect to fit within the bounds of any overflow: hidden ancestors.
+     *
+     * @see https://github.com/dequelabs/axe-core/issues/4253
      */
-    return clientRects.length ? clientRects : [nodeRect];
+    return clientRects.length
+      ? clientRects
+      : filterHiddenRects([nodeRect], overflowHiddenNodes);
```
Comment (straker) [inline lib/commons/dom/get-visible-child-text-rects.js:49]: Decided to leave this in as a failsafe though it technically shouldn't be needed anymore

File: lib/rules/color-contrast-matches.js [hunk 2]
```diff
@@ -147,14 +149,22 @@ function colorContrastMatches(node, virtualNode) {
...
 
-  const rects = range.getClientRects();
-  for (let index = 0; index < rects.length; index++) {
+  const rects = Array.from(range.getClientRects());
+  const overflowNodes = getOverflowHiddenAncestors(virtualNode);
+  return rects.some(rect => {
     //check to see if the rectangle impinges
-    if (visuallyOverlaps(rects[index], node)) {
-      return true;
+    const overlapps = visuallyOverlaps(rect, node);
```
Comment (WilcoFiers) [inline lib/rules/color-contrast-matches.js:156]:
```suggestion
    const overlaps = visuallyOverlaps(rect, node);
```

File: lib/rules/color-contrast-matches.js [hunk 3]
```diff
@@ -147,14 +149,22 @@ function colorContrastMatches(node, virtualNode) {
...
+
+    if (!overflowNodes.length) {
+      return overlapps;
```
Comment (WilcoFiers) [inline lib/rules/color-contrast-matches.js:159]:
```suggestion
      return overlaps;
```

File: lib/rules/color-contrast-matches.js [hunk 4]
```diff
@@ -147,14 +149,22 @@ function colorContrastMatches(node, virtualNode) {
     }
   }
 
-  const rects = range.getClientRects();
-  for (let index = 0; index < rects.length; index++) {
+  const rects = Array.from(range.getClientRects());
+  const overflowNodes = getOverflowHiddenAncestors(virtualNode);
```
Comment (WilcoFiers) [inline lib/rules/color-contrast-matches.js:153]: How about:

```suggestion
  const clippingAncestors = getOverflowHiddenAncestors(virtualNode);
```

**Review by WilcoFiers — CHANGES_REQUESTED** [review]
Editorial

**Review by WilcoFiers — APPROVED** [review]

**Pull request merged**